<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Downstream analysis</title>

<script src="site_libs/header-attrs-2.25/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">climb-jin</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jinp01/CLIMB">
    <span class="fab fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Downstream analysis</h1>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-10-30
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>climb-jin/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20241028code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20241028)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20241028code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20241028)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomjinp01CLIMBtree4b37710cde7dc8645c12d0e438fee16dd773e307targetblank4b37710a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/jinp01/CLIMB/tree/4b37710cde7dc8645c12d0e438fee16dd773e307" target="_blank">4b37710</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomjinp01CLIMBtree4b37710cde7dc8645c12d0e438fee16dd773e307targetblank4b37710a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/jinp01/CLIMB/tree/4b37710cde7dc8645c12d0e438fee16dd773e307" target="_blank">4b37710</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Trash/
    Ignored:    .android/
    Ignored:    Desktop/climb-jin/.DS_Store
    Ignored:    Desktop/climb-jin/analysis/.DS_Store
    Ignored:    Desktop/climb-jin/analysis/pairwise_fitting_cache/
    Ignored:    env_name/
    Ignored:    lemon-x.y.z/

Untracked files:
    Untracked:  .CFUserTextEncoding
    Untracked:  .DS_Store
    Untracked:  .RData
    Untracked:  .RDataTmp
    Untracked:  .Rapp.history
    Untracked:  .Rhistory
    Untracked:  .Xauthority
    Untracked:  .anaconda/
    Untracked:  .anyconnect
    Untracked:  .bash_history
    Untracked:  .bash_profile
    Untracked:  .bash_profile.bak
    Untracked:  .bash_sessions/
    Untracked:  .cache/
    Untracked:  .cisco/
    Untracked:  .conda/
    Untracked:  .condarc
    Untracked:  .config/
    Untracked:  .continuum/
    Untracked:  .cups/
    Untracked:  .dropbox/
    Untracked:  .eclipse/
    Untracked:  .gitconfig
    Untracked:  .ipynb_checkpoints/
    Untracked:  .ipython/
    Untracked:  .julia/
    Untracked:  .juliaup/
    Untracked:  .jupyter/
    Untracked:  .local/
    Untracked:  .matplotlib/
    Untracked:  .oracle_jre_usage/
    Untracked:  .p2/
    Untracked:  .r/
    Untracked:  .snipaste/
    Untracked:  .subversion/
    Untracked:  .tcshrc
    Untracked:  .tooling/
    Untracked:  .virtualenvs/
    Untracked:  .vscode/
    Untracked:  .wns/
    Untracked:  .xonshrc
    Untracked:  .zsh_history
    Untracked:  .zsh_sessions/
    Untracked:  .zshrc
    Untracked:  511hw3.Rmd
    Untracked:  =0.29.21
    Untracked:  =0.4.0
    Untracked:  Applications/
    Untracked:  CLIMB.zip
    Untracked:  Calibre Library/
    Untracked:  Creative Cloud Files Personal Account lydia0425@hotmail.com 3BE21D79656B7BC80A495F8C@AdobeID/
    Untracked:  Creative Cloud Files lydiajin@colostate.edu 5a37fd827490aa859a1a523ab35d0252fdf8d2d8259bfef57b7d4de857d68da6/
    Untracked:  Desktop/.DS_Store
    Untracked:  Desktop/319.pdf
    Untracked:  Desktop/517/
    Untracked:  Desktop/555/
    Untracked:  Desktop/7788/
    Untracked:  Desktop/CLIMB/
    Untracked:  Desktop/File/
    Untracked:  Desktop/Screenshot 2024-10-25 at 09.41.56.png
    Untracked:  Desktop/Screenshot 2024-10-25 at 09.43.11.png
    Untracked:  Desktop/Screenshot 2024-10-25 at 09.43.48.png
    Untracked:  Desktop/Screenshot 2024-10-29 at 10.47.50.png
    Untracked:  Desktop/climb-jin/.Rprofile
    Untracked:  Desktop/climb-jin/.gitattributes
    Untracked:  Desktop/climb-jin/.gitignore
    Untracked:  Desktop/climb-jin/README.md
    Untracked:  Desktop/climb-jin/_workflowr.yml
    Untracked:  Desktop/climb-jin/analysis/about.Rmd
    Untracked:  Desktop/climb-jin/analysis/candidate_latent_classes.Rmd
    Untracked:  Desktop/climb-jin/analysis/license.Rmd
    Untracked:  Desktop/climb-jin/climb-jin.Rproj
    Untracked:  Desktop/climb-jin/code/
    Untracked:  Desktop/climb-jin/data/
    Untracked:  Desktop/climb-jin/output/
    Untracked:  Desktop/climb-jin/pwfits.Rdata
    Untracked:  Desktop/climb-page/
    Untracked:  Desktop/reading/
    Untracked:  Documents
    Untracked:  Downloads
    Untracked:  Dropbox/
    Untracked:  HW5code.Rmd
    Untracked:  HW5code.pdf
    Untracked:  HW6-code.Rmd
    Untracked:  HW6-code.pdf
    Untracked:  IdeaProjects/
    Untracked:  Library/
    Untracked:  MATLAB-Drive/
    Untracked:  Manifest.toml
    Untracked:  Movies/
    Untracked:  Music/
    Untracked:  Pictures/
    Untracked:  Project.toml
    Untracked:  Public/
    Untracked:  PycharmProjects/
    Untracked:  Sigma_chain_py.jld
    Untracked:  Sigma_chain_test.jld
    Untracked:  Sigma_chains_
    Untracked:  Sigma_chains_1
    Untracked:  Sigma_chains_1.csv
    Untracked:  Sigma_chains_10
    Untracked:  Sigma_chains_10.csv
    Untracked:  Sigma_chains_11
    Untracked:  Sigma_chains_11.csv
    Untracked:  Sigma_chains_12
    Untracked:  Sigma_chains_12.csv
    Untracked:  Sigma_chains_13
    Untracked:  Sigma_chains_13.csv
    Untracked:  Sigma_chains_2
    Untracked:  Sigma_chains_2.csv
    Untracked:  Sigma_chains_21.txt
    Untracked:  Sigma_chains_3
    Untracked:  Sigma_chains_3.csv
    Untracked:  Sigma_chains_4
    Untracked:  Sigma_chains_4.csv
    Untracked:  Sigma_chains_423.txt
    Untracked:  Sigma_chains_5
    Untracked:  Sigma_chains_5.csv
    Untracked:  Sigma_chains_6
    Untracked:  Sigma_chains_6.csv
    Untracked:  Sigma_chains_7
    Untracked:  Sigma_chains_7.csv
    Untracked:  Sigma_chains_8
    Untracked:  Sigma_chains_8.csv
    Untracked:  Sigma_chains_9
    Untracked:  Sigma_chains_9.csv
    Untracked:  Sigma_chains_reshaped.txt
    Untracked:  Sigma_mvn_test.csv
    Untracked:  Sigma_star.csv
    Untracked:  Sigma_star_test.csv
    Untracked:  Sigma_star_test_logl.csv
    Untracked:  Sigma_test.csv
    Untracked:  Sites/
    Untracked:  Untitld.Rmd
    Untracked:  Untitld.docx
    Untracked:  Untitled.Rmd
    Untracked:  Untitled.log
    Untracked:  Untitled.tex
    Untracked:  XFeat_training_example.ipynb
    Untracked:  accelerated_features/
    Untracked:  acpt_rate_chain.txt
    Untracked:  anaconda3/
    Untracked:  cgibbs.ipynb
    Untracked:  chain.csv
    Untracked:  chain_test.jld
    Untracked:  climb-page/
    Untracked:  constrained_distributions.ipynb
    Untracked:  consulting.Rmd
    Untracked:  curr_Sigmas_test.csv
    Untracked:  eclipse-workspace/
    Untracked:  eclipse/
    Untracked:  h_mvn_test.csv
    Untracked:  helpers.ipynb
    Untracked:  hyp.RData
    Untracked:  hyp.csv
    Untracked:  hyp.rds
    Untracked:  hyp_Psi0.RData
    Untracked:  hyp_alpha.RData
    Untracked:  hyp_kappa0.RData
    Untracked:  hyp_mu0.RData
    Untracked:  iCloud Drive (Archive)/
    Untracked:  image-matching-webui/
    Untracked:  julia/
    Untracked:  labels.RData
    Untracked:  labels_test.csv
    Untracked:  lgf.txt
    Untracked:  lgf1.txt
    Untracked:  lgf2.txt
    Untracked:  lgf_human.txt
    Untracked:  lgf_human1.txt
    Untracked:  lgf_human2.txt
    Untracked:  ll_mult_curr.csv
    Untracked:  ll_mult_star.csv
    Untracked:  matlab_crash_dump.14732-1
    Untracked:  matlab_crash_dump.17960-1
    Untracked:  matlab_crash_dump.33550-1
    Untracked:  matlab_crash_dump.411-1
    Untracked:  miniconda3/
    Untracked:  minimal_example.ipynb
    Untracked:  mosta.ipynb
    Untracked:  mu0_mvn_test.csv
    Untracked:  mu0_test.csv
    Untracked:  mu_chain_py.jld
    Untracked:  mu_chain_test.csv
    Untracked:  mu_chain_test.jld
    Untracked:  mu_chains.csv
    Untracked:  mu_chains.txt
    Untracked:  mu_chains_1.csv
    Untracked:  mu_chains_10.csv
    Untracked:  mu_chains_11.csv
    Untracked:  mu_chains_12.csv
    Untracked:  mu_chains_13.csv
    Untracked:  mu_chains_2.csv
    Untracked:  mu_chains_3.csv
    Untracked:  mu_chains_3.txt
    Untracked:  mu_chains_4.csv
    Untracked:  mu_chains_5.csv
    Untracked:  mu_chains_6.csv
    Untracked:  mu_chains_7.csv
    Untracked:  mu_chains_8.csv
    Untracked:  mu_chains_9.csv
    Untracked:  mu_chains_reshaped.txt
    Untracked:  mu_hats_test.csv
    Untracked:  occ.csv
    Untracked:  opt/
    Untracked:  out.csv
    Untracked:  plotters.ipynb
    Untracked:  prob_curr.csv
    Untracked:  prob_star.csv
    Untracked:  processors.ipynb
    Untracked:  prop_chain
    Untracked:  prop_chain.csv
    Untracked:  prop_chain.txt
    Untracked:  prop_chain_py.csv
    Untracked:  pwfits.Rdata
    Untracked:  retained_classes.RData
    Untracked:  run_mcmc.ipynb
    Untracked:  run_mcmc_pre.ipynb
    Untracked:  running_mcmc-Copy1.ipynb
    Untracked:  running_mcmc.ipynb
    Untracked:  running_mcmc_Aug6.ipynb
    Untracked:  running_mcmc_Sep2-Copy1.ipynb
    Untracked:  running_mcmc_Sep2.ipynb
    Untracked:  sim.RData
    Untracked:  simdat.RData
    Untracked:  xfeat+lg_torch_hub.ipynb
    Untracked:  xfeat_matching.ipynb
    Untracked:  xfeat_torch_hub.ipynb
    Untracked:  z_chain
    Untracked:  z_chain.csv
    Untracked:  z_chain.txt
    Untracked:  z_chain_py.csv
    Untracked:  z_test
    Untracked:  z_test.csv
    Untracked:  z_test_logl.csv
    Untracked:  zout_test.csv

Unstaged changes:
    Modified:   Desktop/climb-jin/analysis/running_mcmc.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown
(<code>Desktop/climb-jin/analysis/downstream.Rmd</code>) and HTML
(<code>Desktop/climb-jin/docs/downstream.html</code>) files. If you’ve
configured a remote Git repository (see <code>?wflow_git_remote</code>),
click on the hyperlinks in the table below to view the files as they
were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jinp01/CLIMB/blob/4b37710cde7dc8645c12d0e438fee16dd773e307/Desktop/climb-jin/analysis/downstream.Rmd" target="_blank">4b37710</a>
</td>
<td>
Jin Peng
</td>
<td>
2024-10-30
</td>
<td>
Add downstream
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>CLIMB’s output is nothing but the posterior MCMC samples of the
parameters of the normal mixture (means, variance-covariance matrices,
mixing weights, and cluster assignments). These samples can be used for
clustering/classification, visualizations, and statistical inference.
<em>While there are many downstream analyses one could do with such
output</em>, here, we demonstrate implementations of several of the
analyses used in the CLIMB manuscript.</p>
<p>First, let’s load in the MCMC we briefly ran on simulated data in the
<a href="running_mcmc.html">running the MCMC</a> section.</p>
<pre><code>
Attaching package: &#39;dplyr&#39;</code></pre>
<pre><code>The following objects are masked from &#39;package:stats&#39;:

    filter, lag</code></pre>
<pre><code>The following objects are masked from &#39;package:base&#39;:

    intersect, setdiff, setequal, union</code></pre>
<pre class="r"><code>chain &lt;- readRDS(&quot;output/chain.rds&quot;)</code></pre>
<div id="visualizing-convergence" class="section level2">
<h2>Visualizing convergence</h2>
<p>The first thing we should do make some trace plots for some of the
parameters being estimated. This will help us decide when the chain has
entered its stationary distribution. Here, we make plots of a mean,
variance, covariance, and mixing weight.</p>
<pre class="r"><code># The mean in the first dimension of the third cluster
mu31 &lt;- chain$mu_chains[[3]][,1]

# The variance of the first dimension of the second cluster
sigma21 &lt;- chain$Sigma_chains[[2]][1,1,]

# The covariance between dimensions 2 and 3 in the fourth cluster
sigma423 &lt;- chain$Sigma_chains[[4]][2,3,]

# The mixing weight of the seventh cluster
pi7 &lt;- chain$prop_chain[,7]</code></pre>
<p><img src="figure/downstream.Rmd/unnamed-chunk-3-1.png" width="1536" style="display: block; margin: auto;" /></p>
<p>Looking at a few of these trace plots reveals that the chain achieves
stationary around 200 iterations in. Therefore, we discard the first 200
iterations as burn-in, and do our analyses on the remaining 800.</p>
<pre class="r"><code>burnin &lt;- 1:200</code></pre>
</div>
<div id="clustering-dimensions-and-observations" class="section level2">
<h2>Clustering dimensions and observations</h2>
<p>CLIMB can provide a <strong>bi-clustering</strong> of analyzed data.
Row clusters come from the estimated latent classes, of course, though
these classes can even be grouped by their similarity.</p>
<p>CLIMB uses a distance based on class covariances to cluster the
dimensions (columns). The function
<code>compute_distances_between_conditions</code> will compute these
pairwise distances between conditions based on model fit. This was used
in the CLIMB manuscript to determine the relationship among
hematopoietic cell populations based on CTCF binding patterns.</p>
<p>CLIMB’s estimated classes naturally cluster the rows. However, the
function <code>compute_distances_between_clusters</code> will compute
distances between estimated clusters as well. This distance is based on
a symmetrified theoretical Kullback-Liebler divergence.</p>
<pre class="r"><code>par(mfrow = c(1,2))

# The figure isn&#39;t as exciting with simulated data
col_distmat &lt;- compute_distances_between_conditions(chain, burnin)
hc_col &lt;- hclust(as.dist(col_distmat), method = &quot;complete&quot;)
plot(hc_col, xlab = &quot;&quot;, ylab= &quot;&quot;, axes = FALSE, sub = &quot;&quot;, main = &quot;Row cluster dendrogram&quot;)

row_distmat &lt;- compute_distances_between_clusters(chain, burnin)
hc_row &lt;- hclust(as.dist(row_distmat), method = &quot;complete&quot;)
plot(hc_row, xlab = &quot;&quot;, ylab= &quot;&quot;, axes = FALSE, sub = &quot;&quot;, main = &quot;Column cluster dendrogram&quot;)</code></pre>
<p><img src="figure/downstream.Rmd/unnamed-chunk-5-1.png" width="720" style="display: block; margin: auto;" /></p>
<p>These figures aren’t very thrilling on their own, but we can use them
to make some nice heatmaps. CLIMB provides the function
<code>get_row_reordering</code> to facilitate plotting the
bi-clustering:</p>
<pre class="r"><code># Load back in the data
data(&quot;sim&quot;)
dat &lt;- sim$data

# Get a row reordering for plotting
row_reordering &lt;- get_row_reordering(dat = dat, row_clustering = hc_row, chain = chain, burnin = burnin)

# Melt (for ggplot2)
molten1 &lt;-
  dplyr::mutate(dat, &quot;row&quot; = row_reordering) %&gt;%
  tidyr::pivot_longer(cols=!row, names_to = &quot;variable&quot;)

# Rename the factor levels to match the dimensions
levels(molten1$variable) &lt;- as.factor(paste(1:3))

# Plot the bi-clustering heatmap, like in the CLIMB manuscript
p1 &lt;- ggplot(data = molten1,
            aes(x = forcats::fct_relevel(variable, ~ .x[hc_col$order]), # Relevel factors, for column sorting on the plot
                y = row, fill = value)) +
    geom_raster() +
    cowplot::theme_cowplot() +
    theme(legend.position = &quot;none&quot;, axis.ticks.y = element_blank(), axis.text.y = element_blank()) +
    xlab(&quot;&quot;) + ylab(&quot;&quot;) +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = .5)) +
    ggtitle(&quot;Bi-clustering\nheatmap&quot;) +
    scale_fill_gradientn(colours=RColorBrewer::brewer.pal(9,&quot;Greens&quot;))

print(p1)</code></pre>
<p><img src="figure/downstream.Rmd/unnamed-chunk-6-1.png" width="240" style="display: block; margin: auto;" /></p>
</div>
<div id="merging-groups" class="section level2">
<h2>Merging groups</h2>
<p>CLIMB allows you to merge similar groups for the sake of
visualization, for examble on the Genome Browser. In the CLIMB
manuscript, we merged clusters of CTCF binding patterns to color peaks
more in a more parsimonious set of 5 groups on the genome browser. We
show how to do this for an arbitrary number of groups with CLIMB’s
<code>merge_classes</code> function.</p>
<p>The output of a <code>merge_classes</code> call provides the means
and mixing weights of the new parent groups. These means and weights are
aggregated from each groups constituent classes. <code>merged_z</code>
gives the <strong>new group assignments</strong> for each observations.
The <code>clustering</code> obect is simply a call to
<code>stats::hclust()</code>.</p>
<pre class="r"><code># Merge into 5 parent groups
merged &lt;- merge_classes(n_groups = 5, chain = chain, burnin = burnin)

# Check the structure of the merging
str(merged)</code></pre>
<pre><code>List of 6
 $ merged_z    : int [1:1500] 5 1 1 1 1 1 1 1 1 1 ...
 $ merged_mu   : num [1:5, 1:3] 1.19 0 -2.527 -0.632 -2.555 ...
 $ merged_sigma: num [1:3, 1:3, 1:5] 0.8853 0 0.0951 0 1 ...
 $ merged_prop : num [1:5] 0.5797 0.0734 0.1255 0.1043 0.1172
 $ clustering  :List of 7
  ..$ merge      : int [1:7, 1:2] -4 -2 -1 -7 -8 -3 -6 -5 1 2 ...
  ..$ height     : num [1:7] 0.0113 0.022 0.3977 0.7957 0.9305 ...
  ..$ order      : int [1:8] 6 3 8 7 1 2 4 5
  ..$ labels     : chr [1:8] &quot;1&quot; &quot;2&quot; &quot;4&quot; &quot;6&quot; ...
  ..$ method     : chr &quot;average&quot;
  ..$ call       : language hclust(d = as.dist(cluster_dist), method = method)
  ..$ dist.method: NULL
  ..- attr(*, &quot;class&quot;)= chr &quot;hclust&quot;
 $ distmat     : num [1:8, 1:8] 0 0.366 1.638 0.439 0.389 ...
  ..- attr(*, &quot;dimnames&quot;)=List of 2
  .. ..$ : chr [1:8] &quot;1&quot; &quot;2&quot; &quot;4&quot; &quot;6&quot; ...
  .. ..$ : chr [1:8] &quot;1&quot; &quot;2&quot; &quot;4&quot; &quot;6&quot; ...</code></pre>
<p>We can visualize the salient features of the new groups with a simple
visualization of the merged group means:</p>
<pre class="r"><code># For the plot color palette
library(RColorBrewer)

# Melt for plotting
molten2 &lt;- 
  merged$merged_mu %&gt;%
  `colnames&lt;-` (paste0(&quot;cond&quot;, 1:3)) %&gt;%
  as.data.frame() %&gt;%
  dplyr::mutate(group = 1:n()) %&gt;%
  tidyr::pivot_longer(cols = !group, names_to = &quot;condition&quot;, values_to = &quot;mu&quot;)

p2 &lt;- ggplot(data = molten2, aes(group, condition, fill = mu)) +
  geom_tile(color = &quot;black&quot;) +
  scale_fill_distiller(palette = &quot;PiYG&quot;, direction = 1) +
  coord_fixed() +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 0, vjust = 0.5, hjust = 1, size = 18),
          axis.text.y = element_text(size = 20),
          axis.title.x = element_text(size = 18),
          axis.ticks.y = element_blank(),
          axis.ticks.x = element_blank(),
          legend.title = element_text(size = 12))

print(p2)</code></pre>
<p><img src="figure/downstream.Rmd/unnamed-chunk-8-1.png" width="480" style="display: block; margin: auto;" /></p>
</div>
<div id="testing-for-consistency" class="section level2">
<h2>Testing for consistency</h2>
<p>In the CLIMB manuscript, we introduce a statistical test for
consistency of signals based on the <em>partial conjunction
hypothesis</em>:</p>
<p><span class="math display">\[\begin{equation}
  \begin{aligned}
    \mathcal{H}_0^{u/D} &amp;:= \text{less than }u\text{ out of }D\text{
instances of the observed effect are non-null, versus }\\
    \mathcal{H}_1^{u/D} &amp;:= \text{at least }u\text{ out of }D\text{
instances of the observed effect are non-null}
  \end{aligned}
\end{equation}\]</span></p>
<p>where <span class="math inline">\(u\)</span> is some user-defined
consistency threshold, and <span class="math inline">\(D\)</span> is the
dimension of the data. This hypothesis is texting using some combination
of the following quantites:</p>
<p><span class="math display">\[\begin{align}
    &amp;P^{u/D+}:=\sum_{m=1}^M \text{Pr}(\mathbf{x}_i\in h_m \mid
\mathbf{x}) \cdot \mathbf{1} \big[ \sum_{d=1}^D \mathbf{1} (h_{[d]}^m=1)
\geq u\big] \tag{a}\\
    &amp;P^{u/D0} := \sum_{m=1}^M \text{Pr}(\mathbf{x}_i\in h_m \mid
\mathbf{x})\cdot\mathbf{1}\big[\sum_{d=1}^D\mathbf{1}(h_{[d]}^m=0) \geq
u\big] \tag{b} \\
    &amp;P^{u/D-}:=\sum_{m=1}^M \text{Pr}(\mathbf{x}_i\in h_m \mid
\mathbf{x})\cdot\mathbf{1}\big[\sum_{d=1}^D\mathbf{1}(h_{[d]}^m=-1) \geq
u\big] \tag{c} \\
    &amp;P^{u/D}:=\sum_{m=1}^M \text{Pr}(\mathbf{x}_i\in h_m \mid
\mathbf{x})\cdot\mathbf{1}\big[\sum_{d=1}^D\mathbf{1}(h_{[d]}^m \neq 0)
\geq u\big] \tag{d}
\end{align}\]</span></p>
<p>Which quantities you use depends on what you want to test. If you
want to test whether some signals are replicated, and you care about the
sign of the signals, you would employ a replicability test with
quantities (a) and (c). If you wanted to do the same test, but you are
agnostic to the sign of the signal, you would employ quantity (d)
instead. If you wanted to test for consistent behavior (e.g., what is
done in the CLIMB manuscript for differential gene expression along a
lineage), you would employ a test using quantites (a), (b), and (c).</p>
<p>This is all implemented in the CLIMB function
<code>test_consistency</code>. Test consistency takes 6 arguments.</p>
<ul>
<li><code>chain</code>: the output of <code>extract_chains()</code></li>
<li><code>burnin</code>: a vector of iterations (indices) to be used as
burn-in</li>
<li><code>u</code>: the consistency threshold, which is an integer in
<span class="math inline">\(2\ldots D\)</span></li>
<li><code>b</code>: the confidence level in (0,1), corresponding the
<strong>posterior probability that an observation rejects the null
hypothesis</strong>. Here, a natural threshold is 0.5, though larger,
more stringent thresholds could be applied.</li>
<li><code>with_zero</code>: Set this equal to <code>TRUE</code> to
implement the test for consistency across all dimensions, as in the
CLIMB manuscript when testing for differentially expressed genes.</li>
<li><code>agnostic_to_sign</code>: Set equal to <code>TRUE</code> if you
do not care if a signal is in the positive or negative directions.</li>
</ul>
<p>This function will return a boolean vector with <span
class="math inline">\(n\)</span> elements (<span
class="math inline">\(n\)</span> being the sample size). If returns
<code>TRUE</code> when the null is rejected.</p>
<p>So, for example, we can test for which simulated observations are
“consistently expressed” across all 3 dimensions as follows:</p>
<pre class="r"><code>non_differential &lt;- test_consistency(chain, burnin, with_zero = TRUE)
head(non_differential)</code></pre>
<pre><code>[1] FALSE FALSE FALSE  TRUE FALSE FALSE</code></pre>
<p>Here, <code>TRUE</code> sits at the indices of observations who are
likely to be consistent across all 3 dimensions.</p>
<p>Alternatively, we could test for observations which show a consistent
<em>signal</em> (i.e., a -1 or a 1) in at least 2 out of the 3
dimensions. Let’s do this, and also make a precision-recall curve
displaying the results. To make a precision-recall curve, we first need
to know the underlying truth of the simulation.</p>
<pre class="r"><code>#-------------------------------------------------------------------------------
# First, obtain the truth

# Load in true association patterns
data(&quot;true_association&quot;)

# Consistency threshold
u &lt;- 2

# Which indices correspond to consistent signals  at threshold u = 2?
consistent_pos_lab_idx &lt;- apply(true_assoc, 1, function(X) sum(X == 1) &gt;= u)
consistent_neg_lab_idx &lt;- apply(true_assoc, 1, function(X) sum(X == -1) &gt;= u)
consistent_lab_idx &lt;-
      sort(unique(c(
        which(consistent_neg_lab_idx),
        which(consistent_pos_lab_idx)
      )))

# These are the labels corresponding to consistent behavior
consistent_lab_idx</code></pre>
<pre><code>[1] 3 8</code></pre>
<p>Now, we can compute precision and recall across a range of
<code>b</code> in (0,1).</p>
<pre class="r"><code>b_grid &lt;- seq(0, 1, length.out = 1000)

replicable &lt;- test_consistency(chain, burnin, u = 2, b = b_grid)


# Which observations are the replicable ones
rep_idx &lt;- sim$cluster %in% consistent_lab_idx

# Total number of positives in the dataset
P &lt;- sum(rep_idx)

# True positives
TP &lt;- map(replicable, ~ .x &amp; rep_idx)

# False positives
FP &lt;- map(replicable, ~ .x &amp; !rep_idx)

# True positive rate aka recall
recall &lt;- map_dbl(TP, ~ sum(.x) / P)

# precision
precision &lt;- map2_dbl(TP, FP, ~ sum(.x) / (sum(.x) + sum(.y)))

# Plot
df &lt;- data.frame(&quot;recall&quot; = recall, &quot;precision&quot; = precision)
p3 &lt;- ggplot(data = df, aes(x = recall, y = precision)) +
  geom_line() +
  # This horizontal line is the true proportion of replicable observations
  geom_hline(yintercept = mean(rep_idx), lty = 2, color = &quot;darkcyan&quot;) +
  lims(x = c(0,1), y = c(0,1)) +
  theme_minimal()
  
p3</code></pre>
<p><img src="figure/downstream.Rmd/unnamed-chunk-11-1.png" width="384" style="display: block; margin: auto;" /></p>
</div>
<div id="multiple-chains" class="section level2">
<h2>Multiple chains</h2>
<p>It is possible that one might want to do downstream analysis on
multiple outputs from different CLIMB analyses. For example, you may be
using CLIMB to analyze ChIP- or ATAC-seq data genome-wide. Due to
computational considerations, you may have analyzed each chromosome
separately. Then, you will have ~20 different MCMC chains that you could
consider together for downstream analysis. CLIMB facilitates analyses
with the above functions with the addition of the
<code>multichain</code> argument. For the functions</p>
<ul>
<li><code>compute_distances_between_conditions</code></li>
<li><code>compute_distances_between_clusters</code></li>
<li><code>get_row_reordering</code></li>
<li><code>merge_classes</code></li>
</ul>
<p>if the argument <code>multichain</code> is set to <code>TRUE</code>,
then CLIMB will expect that you pass a <em>list</em> of MCMC outputs,
rather than just 1. A list of burn-in values for each chain can be
supplied, or 1 set of burn-in can be applied uniformly across all
chains.</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.3.2 (2023-10-31)
Platform: x86_64-apple-darwin20 (64-bit)
Running under: macOS Sonoma 14.5

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/New_York
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] RColorBrewer_1.1-3 ggplot2_3.5.0      purrr_1.0.2        CLIMB_1.1.0       
[5] dplyr_1.1.4        workflowr_1.7.1   

loaded via a namespace (and not attached):
 [1] gtable_0.3.4         xfun_0.42            bslib_0.6.1         
 [4] processx_3.8.3       callr_3.7.5          tzdb_0.4.0          
 [7] vctrs_0.6.5          tools_4.3.2          ps_1.7.6            
[10] generics_0.1.3       parallel_4.3.2       tibble_3.2.1        
[13] fansi_1.0.6          highr_0.10           pkgconfig_2.0.3     
[16] lifecycle_1.0.4      farver_2.1.1         compiler_4.3.2      
[19] stringr_1.5.1        git2r_0.33.0         brio_1.1.4          
[22] munsell_0.5.0        getPass_0.2-4        codetools_0.2-19    
[25] httpuv_1.6.14        htmltools_0.5.7      sass_0.4.8          
[28] yaml_2.3.8           later_1.3.2          pillar_1.9.0        
[31] jquerylib_0.1.4      whisker_0.4.1        tidyr_1.3.1         
[34] cachem_1.0.8         iterators_1.0.14     JuliaCall_0.17.5    
[37] abind_1.4-5          foreach_1.5.2        tidyselect_1.2.0    
[40] digest_0.6.34        mvtnorm_1.2-4        stringi_1.8.3       
[43] LaplacesDemon_16.1.6 labeling_0.4.3       forcats_1.0.0       
[46] cowplot_1.1.3        rprojroot_2.0.4      fastmap_1.1.1       
[49] grid_4.3.2           colorspace_2.1-0     cli_3.6.2           
[52] magrittr_2.0.3       utf8_1.2.4           readr_2.1.5         
[55] withr_3.0.0          scales_1.3.0         promises_1.2.1      
[58] rmarkdown_2.25       httr_1.4.7           hms_1.1.3           
[61] evaluate_0.23        knitr_1.45           testthat_3.2.1      
[64] doParallel_1.0.17    rlang_1.1.3          Rcpp_1.0.12         
[67] glue_1.7.0           rstudioapi_0.15.0    jsonlite_1.8.8      
[70] R6_2.5.1             plyr_1.8.9           fs_1.6.3            </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
